sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -318,7 +318,10 @@ class Analyzer(
         gid: Expression): Expression = {
       expr transform {
         case e: GroupingID =>
-          if (e.groupByExprs.isEmpty || e.groupByExprs == groupByExprs) {
+          def sameExpressions(e1: Seq[Expression], e2: Seq[Expression]): Boolean = {
```

Comment:
Is this a bug not related to this PR?

Reply:
I think so, basically we should always use `semanticEquals` when matching expressions.

Reply:
Anyway my PR exposed this bug as now `Dataset.col` returns a slightly different attribute with a metadata.